specs/altair/sync-protocol.md

```diff
@@ -256,3 +351,395 @@ def process_light_client_update(store: LightClientStore,
+                                           optimistic_update: OptimisticLightClientUpdate,
+                                           current_slot: Slot,
+                                           genesis_validators_root: Root) -> None:
+    process_light_client_update(
+        store,
+        LightClientUpdate(
+            attested_header=optimistic_update.attested_header,
+            next_sync_committee=SyncCommittee(),
+            next_sync_committee_branch=[Bytes32() for _ in range(floorlog2(NEXT_SYNC_COMMITTEE_INDEX))],
+            finalized_header=BeaconBlockHeader(),
+            finality_branch=[Bytes32() for _ in range(floorlog2(FINALIZED_ROOT_INDEX))],
+            sync_aggregate=optimistic_update.sync_aggregate,
+            fork_version=optimistic_update.fork_version),
+        current_slot,
+        genesis_validators_root)
+```
+
+## Networking
+
+This section extends the [networking specification for Altair](./p2p-interface.md) with additional messages, topics and data to the Req-Resp and Gossip domains.
+
+### Configuration
+
+| Name | Value | Description |
+| - | - | - |
+| `MAX_REQUEST_LIGHT_CLIENT_UPDATES` | `2**7` (= 128) | Maximum number of `LightClientUpdate` instances in a single request |
+
+### The gossip domain: gossipsub
+
+A gossip mesh is added to allow light clients to stay in sync with the network.
+
+#### Topics and messages
+
+A new global topic is added to provide light clients with the latest updates for
+staying in sync with the network.
+
+| name | Message Type |
+| - | - |
+| `optimistic_light_client_update` | `OptimisticLightClientUpdate` |
+
+##### Global topics
+
+###### `optimistic_light_client_update`
+
+This topic is used to propagate the latest `OptimisticLightClientUpdate` to light
+clients, allowing them to stay in sync with the network.
+
+The following validations MUST pass before forwarding the `optimistic_update`
+on the network.
+- _[IGNORE]_ The optimistic update is not attesting to the latest block's parent block --
+  i.e. validate that `hash_tree_root(optimistic_update.attested_header) == state.latest_block_header.parent_root`.
+- _[REJECT]_ The optimistic update has an unexpected fork version --
+  i.e. validate that `optimistic_update.fork_version == state.fork.current_version`
+- _[REJECT]_ The optimistic update sync aggregate differs from the one included in the latest block --
+  i.e. with `block` referring to the latest block,
+  validate that `optimistic_update.sync_aggregate == block.message.body.sync_aggregate`
+
+Note that the `sync_aggregate` has already been validated when processing the
+latest block. While different valid aggregate signatures may be formed,
+only a single canonical one (the same one that was included in the block) should
+be propagated across the network.
+
+### The Req/Resp domain
+
+#### Messages
+
+##### LightClientUpdatesByRange
+
+**Protocol ID:** `/eth2/beacon_chain/req/light_client_updates_by_range/1/`
```

Comment:
I would suggest that we consider request by some kind of *id* instead of range. The reason is that two folds:
1. The networking allows requesters to ask specific updates. If the network has competing forks, all clients could still update based on their trusting sync-comm. In the by-range approach, if there is big fraction of network answers a client's request for a particular period with the wrong sync-comm, the client could not easily find an update that are unique to its trusted current-sync-comm. 
2. Requesting data object by key-value will translate well to other DHT-like p2p network. Specifically, portal network will work better with a key-value mechanism.

Reply:
Thanks, that could indeed be a problem when there is non-finality for more than 2 sync committee periods (around 2 days). If non-finality is shorter, the sync committees are still stable (`current_sync_committee` and `next_sync_committee` are fixed at the start of a period), so any update from any of the forks should continue to apply fine. 

On the server side, `is_better_light_client_update` follows the majority choice picked by the sync committee in that case. Once finality is restored, this endpoint would return stable data once more. 

On the client side new updates are only applied if a sufficient part of the sync committee signed them and a long timeout has elapsed.

I think that with heavy forking across multiple sync committee periods, that you will end up having to re-sync a light client from a trusted checkpoint eventually. As in, that this is a problem not introduced by the proposed network layer, but a pre-existing conceptual limitation. If this proves to be a network layer problem, an additional endpoint taking both the period and the `hash_tree_root(sync_committee)` could help solve this. This would be in line with the duality of `BeaconBlocksByRange` (for finalized data) and `BeaconBlocksByRoot` (for non-finalized data).

Regarding key-value for portal network, the key here would just be the sync committee period. The value would change when a better update is found within a sync committee period (`is_better_light_client_update`).

Reply:
Regarding changing the request to be keyed by some kind of _id_ instead of by `SyncCommitteePeriod`, I am not sure if it would actually help with (1) allowing client's to stay in sync with their view of the network.

The proposed protocol depends on the light client syncing from a trusted checkpoint within the weak subjectivity period. Doing that guarantees that it is not possible for multiple finalized chains to co-exist. This means, that up until the finalized checkpoint, there is only 1 possible fork that is propagated through the network, so using the `SyncCommitteePeriod` as key should be fine, also allowing to sync with the least amount of req/rsp round trips.

For the non-finalized slots, multiple forks could exist. However, if in a given sync committee period, any block has been finalized, the current and the next sync committee are fixed. This means that regardless of forking within the current and next sync committee period, light clients will be able to follow along.

If non-finality lasts through an entire sync committee period, though, the situation that you describe may occur. Multiple valid sync committees may be active, and it may be difficult for a light client to follow the specific fork that it is locked on-to. In that situation, having a request that is keyed by some kind of _id_ for the specific fork would indeed allow the light client to be more specific in its request. However, the light client will still have trouble discovering what peers are actually following that specific fork, ultimately boiling down to attempting to request updates for the specific fork and see if there is any response, vs just attempting to request any updates for the given period and see whether it is from the specific fork. Also, if the light client ends up picking an incorrect fork, even when it succeeded following its specific fork more efficiently, it would still not be able to recover if a different fork ends up being finalized, requiring a re-sync from a trusted checkpoint. Once finality is eventually restored, there is again only 1 possible fork being propagated through the network up until the new finalized checkpoint, enabling a deterministic re-sync.

Note that the situation should only occur if a majority of the sync committee originally signed a fork that becomes unviable in a later sync committee period. If they signed the dominating fork instead, even in extended periods of non-finality, light clients should end up picking the correct fork.

Ultimately, I think that following non-finalized parts of the chain remains a gamble of picking the correct fork, regardless of how the requests are keyed. If the correct fork is picked, it will eventually become easy again to follow it. If an incorrect one is picked, it will result in having to re-sync from a trusted checkpoint.

As for (2) the portal network, I guess using the `SyncCommitteePeriod` as an ID should be fine, or alternatively, a combination of the `SyncCommitteePeriod` with `hash_tree_root(sync_committee)`. However, even there, once finality is restored, the unviable forks would eventually end up being abandoned, resulting in light clients having to re-sync.

Reply:
One more clarification on the addressing by some kind of _id_ instead of just sync committee period.

To implement the server side correctly, I agree that it **is** necessary to internally keep track of the _id_ so that the best `LightClientUpdate` avoids referring to an orphaned block in its `active_header`. My proposal to avoid exposing the root is solely related to the proposed libp2p network protocol.

From `Nimbus-eth2` server, these are the internally tracked updates:
```
    bestLightClientUpdates*:
       Table[SyncCommitteePeriod, altair.LightClientUpdate]
       ## Stores the `LightClientUpdate` with the most `sync_committee_bits` per
       ## `SyncCommitteePeriod`. Updates with a finality proof have precedence.

     pendingBestLightClientUpdates*:
       Table[(SyncCommitteePeriod, Eth2Digest), altair.LightClientUpdate]
       ## Same as `bestLightClientUpdates`, but for `SyncCommitteePeriod` with
       ## `next_sync_committee` that are not finalized. Key is `(period,
       ## hash_tree_root(current_sync_committee | next_sync_committee)`.

     latestLightClientUpdate*: altair.LightClientUpdate
       ## Tracks the `LightClientUpdate` for the latest slot. This may be older
       ## than head for empty slots or if not signed by sync committee.

     optimisticLightClientUpdate*: OptimisticLightClientUpdate
       ## Tracks the `OptimisticLightClientUpdate` for the latest slot. This may
       ## be older than head for empty slots or if not signed by sync committee.
```

Whenever fork choice selects a new head, `bestLightClientUpdates` is updated with the corresponding `pendingBestLightClientUpdates`. As soon as any block within a sync committee period finalizes, `active_header` is guaranteed to always be on the correct chain, so the forks are no longer tracked independently.